pkg/topology/kademlia/kademlia.go

```diff
@@ -668,35 +672,39 @@ func (k *Kad) connect(ctx context.Context, peer swarm.Address, ma ma.Multiaddr,
 func (k *Kad) Announce(ctx context.Context, peer swarm.Address) error {
 	addrs := []swarm.Address{}
 
-	_ = k.connectedPeers.EachBinRev(func(connectedPeer swarm.Address, _ uint8) (bool, bool, error) {
-		if connectedPeer.Equal(peer) {
-			return false, false, nil
+	//r := rand.New(securerand.NewSource())
+
+	for bin := uint8(0); bin < swarm.MaxBins; bin++ {
+
+		connectedPeers, err := randomSubset(k.connectedPeers.BinPeers(bin), broadcastBinSize)
+		if err != nil {
+			return err
 		}
 
-		addrs = append(addrs, connectedPeer)
-
-		// this needs to be in a separate goroutine since a peer we are gossipping to might
-		// be slow and since this function is called with the same context from kademlia connect
-		// function, this might result in the unfortunate situation where we end up on
-		// `err := k.discovery.BroadcastPeers(ctx, peer, addrs...)` with an already expired context
-		// indicating falsely, that the peer connection has timed out.
-		k.wg.Add(1)
-		go func(connectedPeer swarm.Address) {
-			defer k.wg.Done()
-			if err := k.discovery.BroadcastPeers(context.Background(), connectedPeer, peer); err != nil {
-				k.logger.Debugf("could not gossip peer %s to peer %s: %v", peer, connectedPeer, err)
+		for _, connectedPeer := range connectedPeers {
+			if connectedPeer.Equal(peer) {
+				continue
 			}
-		}(connectedPeer)
 
-		return false, false, nil
-	})
+			addrs = append(addrs, connectedPeer)
+
+			k.wg.Add(1)
+			go func(connectedPeer swarm.Address) {
+				defer k.wg.Done()
+				if err := k.discovery.BroadcastPeers(context.Background(), connectedPeer, peer); err != nil {
+					k.logger.Debugf("could not gossip peer %s to peer %s: %v", peer, connectedPeer, err)
+				}
+			}(connectedPeer)
+		}
+	}
 
 	if len(addrs) == 0 {
 		return nil
 	}
 
 	err := k.discovery.BroadcastPeers(ctx, peer, addrs...)
 	if err != nil {
+		fmt.Println("DISCONNECT")
```

Comment:
do we want to keep this line or was it intended to be something temp?

Reply:
oh snap...